### PR TITLE
Improve README; add benchmarks; add QOL functions

### PR DESCRIPTION
**Pull request by ericmillin**
Added examples to the README. As a part of that, I added a few factory functions to make things cleaner. I also added some benchmarks.

### CODE REVIEW

File: fluent/protocol/chunk.go [hunk 1]
```diff
@@ -10,6 +10,15 @@ import (
 	"github.com/tinylib/msgp/msgp"
 )
 
+// ChunkEncoder wraps methods to encode a message and generate
+// "chunk" IDs for use with Fluent's chunk-ack protocol. See
+// https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#response
+// for more information.
+type ChunkEncoder interface {
```
Comment (ericmillin) [inline fluent/protocol/chunk.go:17]: moved from another file

File: fluent/client/client.go [hunk 2]
```diff
@@ -88,22 +132,11 @@ func (c *Client) Connect() error {
...
-	// if c.Session != nil {
-	// 	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", c.Session.Hostname,
-	// 		c.Session.Port), t)
-	// 	if err != nil {
-	// 		return err
-	// 	}
-	// 	c.Session.Connection = conn
-	// }
-	return nil
+	if c.Session != nil {
```
Comment (ericmillin) [inline fluent/client/client.go:135]: This is bugged. Will remove along with example in  README

File: cmd/bm/fluent_forward_go/bm_test.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+	c := &client.Client{
+		Timeout: 3 * time.Second,
+		ConnectionFactory: &client.TCPConnectionFactory{
+			Target: client.ServerAddress{
+				Hostname: "localhost",
+				Port:     24224,
+			},
+		},
+	}
```
Comment (BernhardSchmid) [inline cmd/bm/fluent_forward_go/bm_test.go:24]: Why not using our own stuff?

```
c = client.New(client.ConnectionOptions{})
c.Timeout = 3 * time.Second
```

If changing here also change the occurrences below

**Review by gufranmirza — APPROVED** [review]
LGTM

File: fluent/protocol/message.go [hunk 4]
```diff
@@ -135,6 +153,22 @@ type MessageExt struct {
 	Options   *MessageOptions
 }
 
+// NewMessage creates a Message from the supplied
+// tag and record. It will set Timestamp to
+// time.Now().UTC().Unix().
```
Comment (BernhardSchmid) [inline fluent/protocol/message.go:158]: Should be: `time.Now().UTC()` 
 ... or `EventTimeNow()`

File: fluent/protocol/message.go [hunk 5]
```diff
@@ -21,6 +23,22 @@ type Message struct {
...
+func NewMessage(
+	tag string,
+	record Record,
+) *Message {
+	msg := &Message{
+		Tag:       tag,
+		Timestamp: time.Now().UTC().Unix(),
```
Comment (BernhardSchmid) [inline fluent/protocol/message.go:35]: is type `int64` for field Timestamp correct? Shouldn't it be `protocol.EventTime`?

This does belong to the `Message` struct, but I can't remark there.

**Review by BernhardSchmid — COMMENTED** [review]
I'm used to run `ginkgo -r` in the home directory of the repo. With this changes it start with the following error messages - probably due to the benchmark tests.

```
fluent-forward-go> ginkgo -r
PASS
testing: warning: no tests to run
Found no test suites, did you forget to run "ginkgo bootstrap"?PASS
testing: warning: no tests to run
Found no test suites, did you forget to run "ginkgo bootstrap"?[1639752731] Fluent Suite - 0/0 specs  SUCCESS! 306.259µs PASS
```

File: fluent/protocol/message.go (hunk 4 above)
Comment (ericmillin) [reply fluent/protocol/message.go:158]: This is correct for this type. MessageExt takes EventTime. The difference is part of the spec

**Pull request merged**